File: python/longest_increasing_path_in_a_matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: list[list[int]]) -> int:
        """time complexity: O(n * m)"""
        cache: dict[tuple[int, int], int] = {}
        for r in range(len(matrix)):
            for c in range(len(matrix[r])):
                self.dfs(r, c, cache, -1, matrix)
        return max(cache.values())

    def dfs(
        self,
        row: int,
        col: int,
        cache: dict[tuple[int, int], int],
        prev: int,
        matrix: list[list[int]],
    ) -> int:
        if (
            row < 0
            or row >= len(matrix)
            or col < 0
            or col >= len(matrix[row])
            or matrix[row][col] <= prev
        ):
            return 0
        if (row, col) in cache:
            return cache[(row, col)]
        cache[(row, col)] = max(
            1 + self.dfs(row - 1, col, cache, matrix[row][col], matrix),
            1 + self.dfs(row + 1, col, cache, matrix[row][col], matrix),
            1 + self.dfs(row, col - 1, cache, matrix[row][col], matrix),
            1 + self.dfs(row, col + 1, cache, matrix[row][col], matrix),
        )
        return cache[(row, col)]


def test_longestIncreasingPath_case_1():
    # arrange
    matrix: list[list[int]] = [[9, 9, 4], [6, 6, 8], [2, 1, 1]]
    expected: int = 4

    # act
    actual = Solution().longestIncreasingPath(matrix)

    # assert
    assert expected == actual


def test_longestIncreasingPath_case_2():
    # arrange
    matrix: list[list[int]] = [[3, 4, 5], [3, 2, 6], [2, 2, 1]]
    expected: int = 4

    # act
    actual = Solution().longestIncreasingPath(matrix)

    # assert
    assert expected == actual


def test_longestIncreasingPath_case_3():
    # arrange
    matrix: list[list[int]] = [[1]]
    expected: int = 1

    # act
    actual = Solution().longestIncreasingPath(matrix)

    # assert
    assert expected == actual

File: rust/src/code/longest_increasing_path_in_a_matrix.rs
use std::cmp::max;
use std::collections::HashMap;

struct Solution;

impl Solution {
    // time complexity: O(m * n)
    pub fn longest_increasing_path(matrix: Vec<Vec<i32>>) -> i32 {
        let mut cache: HashMap<(usize, usize), i32> = HashMap::new();
        for r in 0..matrix.len() {
            for c in 0..matrix[r].len() {
                Self::dfs(r, c, &mut cache, -1, &matrix);
            }
        }
        let result: Option<&i32> = cache.values().max();
        match result {
            Some(v) => *v,
            None => 0,
        }
    }

    fn dfs(
        r: usize,
        c: usize,
        cache: &mut HashMap<(usize, usize), i32>,
        prev: i32,
        matrix: &Vec<Vec<i32>>,
    ) -> i32 {
        if r >= matrix.len() || c >= matrix[r].len() || matrix[r][c] <= prev {
            return 0;
        }
        if cache.contains_key(&(r, c)) {
            return cache[&(r, c)];
        }
        let mut up: i32 = 1;
        if r.checked_sub(1).is_some() {
            up = 1 + Self::dfs(r - 1, c, cache, matrix[r][c], matrix);
        }
        let down: i32 = 1 + Self::dfs(r + 1, c, cache, matrix[r][c], matrix);
        let mut left: i32 = 1;
        if c.checked_sub(1).is_some() {
            left = 1 + Self::dfs(r, c - 1, cache, matrix[r][c], matrix);
        }
        let right: i32 = 1 + Self::dfs(r, c + 1, cache, matrix[r][c], matrix);
        let result: i32 = max(up, max(down, max(left, right)));
        cache.insert((r, c), result);
        result
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test_longest_increasing_path_case_1() {
        // arrange
        let matrix: Vec<Vec<i32>> = vec![vec![9, 9, 4], vec![6, 6, 8], vec![2, 1, 1]];
        let expected: i32 = 4;

        // act
        let actual = Solution::longest_increasing_path(matrix);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_longest_increasing_path_case_2() {
        // arrange
        let matrix: Vec<Vec<i32>> = vec![vec![3, 4, 5], vec![3, 2, 6], vec![2, 2, 1]];
        let expected: i32 = 4;

        // act
        let actual = Solution::longest_increasing_path(matrix);

        // assert
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_longest_increasing_path_case_3() {
        // arrange
        let matrix: Vec<Vec<i32>> = vec![vec![1]];
        let expected: i32 = 1;

        // act
        let actual = Solution::longest_increasing_path(matrix);

        // assert
        assert_eq!(actual, expected);
    }
}
